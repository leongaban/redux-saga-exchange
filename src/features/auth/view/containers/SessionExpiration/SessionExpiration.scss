@import "~shared/view/styles/mixins";

.session-expiration {
  width: 20.2857rem;

  &__clock {
    height: 5.1429rem;
    margin: 0 auto;
    margin-top: 2.4286rem;
    background: url('./img/clock.svg') no-repeat;
    background-position: center;
  }

  &__title {
    margin-top: 2.0714rem;
    text-align: center;
    font-size: 0.7857rem;

    @include themed('color', 'main-text-color');
  }

  &__timer {
    display: flex;
    margin-top: 1.1429rem;
    font-size: 2.2857rem;
    font-weight: 300;
    justify-content: center;

    @include themed('color', 'main-text-color');
  }

  &__button {
    width: 15rem;
    padding: 1.3rem 0;
    margin: auto;
  }
}
