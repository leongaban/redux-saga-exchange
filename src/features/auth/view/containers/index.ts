export { default as LoginForm } from './LoginForm/LoginForm';
export { default as RegistrationForm } from './RegistrationForm/RegistrationForm';
export { default as PasswordRecoveryForm } from './PasswordRecoveryForm/PasswordRecoveryForm';
export { default as ChangePasswordForm } from './ChangePasswordForm/ChangePasswordForm';
export { default as SessionExpiration } from './SessionExpiration/SessionExpiration';
export { default as ConfirmEmail } from './ConfirmEmail/ConfirmEmail';
export { default as TwoFactorForm } from './TwoFactorForm/TwoFactorForm';
export { default as Logout } from './Logout/Logout';
export { default as TermsOfService } from './TermsOfService/TermsOfService';
export { default as TwoFactorNotice } from './TwoFactorNotice/TwoFactorNotice';
export { default as AuthLayout } from './AuthLayout/AuthLayout';
export { default as MAuthLayout } from './AuthLayout/mobile/MAuthLayout';
export { default as SSLNotice } from './SSLNotice/SSLNotice';
