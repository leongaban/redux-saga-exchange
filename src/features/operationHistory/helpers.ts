import * as R from 'ramda';

import { IOperation } from 'shared/types/models';
import { sortArray } from 'shared/helpers/sort';

export function applyOperationHistoryDiff(
  currentOperations: IOperation[], operationsDiff: IOperation[],
): IOperation[] {
  const updatedOperations = operationsDiff.reduce((curr, diff) => {
    const matchIndex = curr.findIndex(x => diff.id === x.id);
    return matchIndex !== -1 ? R.update(matchIndex, diff, curr) : R.append(diff, curr);
  }, currentOperations);
  return sortArray(updatedOperations, {
    column: 'creationDate',
    direction: 'descend',
    kind: 'date',
  });
}
