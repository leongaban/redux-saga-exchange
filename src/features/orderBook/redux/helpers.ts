import { Decimal } from 'decimal.js';

import { ITradeOrder, IActiveOrder, IOrderBookOrder } from 'shared/types/models';

export function makeDecimalGrouper(
  decimals: number, round: (x: number, decimals: number) => Decimal,
) {
  return function groupByDecimals(tradeOrders: ITradeOrder[], activeOrders: IActiveOrder[]): IOrderBookOrder[] {
    if (tradeOrders.length === 0) {
      return [];
    }

    const checkIfGroupHasMyOrder = (current: Decimal) => {
      return !!activeOrders.find(activeOrder => {
        return round(activeOrder.limitPrice, decimals).equals(current);
      });
    };

    let currentGroup = round(tradeOrders[0].price, decimals);
    let currentGroupVolume = tradeOrders[0].volume;
    let currentGroupTotal = tradeOrders[0].total;
    const acc: IOrderBookOrder[] = [];

    tradeOrders.slice(1).forEach(x => {
      const nextGroup = round(x.price, decimals);

      if (nextGroup.equals(currentGroup)) {
        currentGroupVolume += x.volume;
        currentGroupTotal += x.total;
      } else {
        acc.push({
          price: currentGroup.toNumber(),
          volume: currentGroupVolume,
          total: currentGroupTotal,
          isMine: checkIfGroupHasMyOrder(currentGroup),
        });
        currentGroup = nextGroup;
        currentGroupVolume = x.volume;
        currentGroupTotal = x.total;
      }
    });

    if (currentGroupVolume > 0) {
      acc.push({
        price: currentGroup.toNumber(),
        volume: currentGroupVolume,
        total: currentGroupTotal,
        isMine: checkIfGroupHasMyOrder(currentGroup),
      });
    }

    return acc;
  };
}
