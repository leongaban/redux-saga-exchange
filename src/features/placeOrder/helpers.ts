import { Decimal } from 'decimal.js';

import { ITradeOrder, OrderSide } from 'shared/types/models';

import * as NS from './namespace';

function calculateMarketOrderVolume(arr: ITradeOrder[], balance: number, resultVolume: number): number {
  if (arr.length === 0) {
    return resultVolume;
  }
  const { price, volume } = arr[0];
  const total = price * volume;
  if (total > balance) {
    return price === 0 ? resultVolume : resultVolume + balance / price;
  }

  return calculateMarketOrderVolume(arr.slice(1), balance - total, resultVolume + volume);
}

export function calculateVolume({
  formType,
  balancePercentage,
  baseCurrencyBalance,
  counterCurrencyBalance,
  orderType,
  askOrders,
  price,
}: NS.ICalculateVolumeOptions) {
  const volumeValue = (() => {
    switch (formType) {
      case 'sell':
        const portionOfBaseBalance = Decimal.mul(baseCurrencyBalance, balancePercentage).div(100).toNumber();

        return balancePercentage > 0
          ? portionOfBaseBalance
          : 0;

      case 'buy':
        const portionOfCounterBalance = Decimal.mul(counterCurrencyBalance, balancePercentage).div(100).toNumber();
        const normalizedPrice = isNaN(Number(price)) ? 0 : Number(price);
        const buyVolume = orderType === 'Market'
          ? calculateMarketOrderVolume(askOrders, portionOfCounterBalance, 0)
          : normalizedPrice === 0 ? 0 : portionOfCounterBalance / normalizedPrice;
        return buyVolume;
    }
  })();

  return volumeValue;
}

export function calculateTotal(volume?: string, price?: string) {
  if (volume !== undefined && price !== undefined) {
    return +volume * +price;
  }
  return void 0;
}

export function shouldShowLimitOrderPriceWarning(price: number, currentMarketPrice: number, formType: OrderSide) {
  const diffProportion = (price - currentMarketPrice) / currentMarketPrice;
  switch (formType) {
    // 0.05 === 5 percent
    case 'buy':
      return diffProportion >= 0.05;
    case 'sell':
      return diffProportion <= -0.05;
    default:
      return false;
  }
}
