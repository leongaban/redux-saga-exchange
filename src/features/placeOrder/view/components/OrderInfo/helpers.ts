import * as R from 'ramda';
import { ITradeOrder, ICurrencyPair } from 'shared/types/models';
import { PurchasedAssetVolume, FeeType } from '../../../namespace';

export function calculateOrderFee(feeType: FeeType, purchasedAssetVolume: number, currencyPair: ICurrencyPair) {
  const { takerFeeLimit, makerFeeLimit, takerFee, makerFee } = currencyPair;

  const currencyFee = feeType === 'taker' ? takerFee : makerFee;
  const feeLimit = feeType === 'taker' ? takerFeeLimit : makerFeeLimit;

  return R.max(feeLimit, purchasedAssetVolume * currencyFee);
}

export function getBuyFormPurchasedAssetVolume(
  takerOrders: ITradeOrder[], placedOrderVolume: number,
): PurchasedAssetVolume {
  if (takerOrders.length === 0) {
    // no taker orders = whole fee is maker
    return { taker: 0, maker: placedOrderVolume };
  } else {
    const takerOrdersVolume = R.sum(takerOrders.map(x => x.volume));
    return placedOrderVolume > takerOrdersVolume
      // current book orders are not enough and a new order entry will be added
      ? { taker: takerOrdersVolume, maker: placedOrderVolume - takerOrdersVolume }
      // current book orders are enough and a new order entry won't be added
      : { taker: placedOrderVolume, maker: 0 };
  }
}

export function getSellFormPurchasedAssetVolume(
  takerOrders: ITradeOrder[], placedOrderPrice: number, placedOrderVolume: number,
): PurchasedAssetVolume {
  const placedOrderTotal = placedOrderPrice * placedOrderVolume;
  if (takerOrders.length === 0) {
    // no taker orders = whole fee is maker
    return { taker: 0, maker: placedOrderTotal };
  } else {
    const takerOrdersTotal = R.sum(takerOrders.map(x => x.total));
    return placedOrderTotal > takerOrdersTotal
      // current book orders are not enough and a new order entry will be added
      ? { taker: takerOrdersTotal, maker: placedOrderTotal - takerOrdersTotal }
      // current book orders are enough and a new order entry won't be added
      : { taker: placedOrderTotal, maker: 0 };
  }
}
