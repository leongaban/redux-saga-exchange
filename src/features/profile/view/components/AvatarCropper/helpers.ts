import { IImageFile } from 'shared/types/models';

export function createImageFile(
  name: string, preview: string, blobParts: Blob[], blobOptions?: BlobPropertyBag,
): IImageFile {
  const img = new Blob(blobParts, blobOptions) as IImageFile;
  img.name = name;
  img.preview = preview;
  return img;
}
