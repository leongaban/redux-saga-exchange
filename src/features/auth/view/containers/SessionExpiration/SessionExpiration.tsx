import * as React from 'react';
import block from 'bem-cn';
import { connect, Dispatch } from 'react-redux';
import { bindActionCreators } from 'redux';
import { bind } from 'decko';

import { ITranslateProps, i18nConnect } from 'services/i18n';
import { IAppReduxState } from 'shared/types/app';
import { Button, Modal } from 'shared/view/elements';
import { sessionExpirationLimit } from 'shared/constants';

import moment from 'services/moment';
import { actions, selectors } from '../../../redux';
import './SessionExpiration.scss';

interface IOwnProps {
  onClose(): void;
  onSessionExpired(): void;
}

interface IStateProps {
  isTimerStarted: boolean;
  timerValue: number;
}

interface IDispatchProps {
  startTimer: typeof actions.startTimer;
  stopTimer: typeof actions.stopTimer;
}

type IProps = IStateProps & IDispatchProps & ITranslateProps & IOwnProps;

function mapState(state: IAppReduxState): IStateProps {
  return {
    isTimerStarted: selectors.selectIsTimerStarted(state),
    timerValue: selectors.selectTimerValue(state),
  };
}

function mapDispatch(dispatch: Dispatch<any>): IDispatchProps {
  return bindActionCreators({
    startTimer: actions.startTimer,
    stopTimer: actions.stopTimer,
  }, dispatch);
}

const b = block('session-expiration');

class SessionExpiration extends React.Component<IProps> {

  public componentDidMount() {
    const { startTimer } = this.props;
    startTimer(sessionExpirationLimit);
  }

  public componentDidUpdate() {
    const { timerValue, onSessionExpired, onClose } = this.props;
    if (timerValue === 0) {
      onClose();
      onSessionExpired();
    }
  }

  public render() {
    const { isTimerStarted, timerValue, translate: t } = this.props;
    const minutes = Math.floor(timerValue / 60);
    const seconds = Math.floor(timerValue % 60);
    const minutesFormatted = moment(minutes, 'minutes').format('mm');
    const secondsFormatted = moment(seconds, 'seconds').format('ss');
    return (
      <Modal
        title="session expiration"
        isOpen={isTimerStarted}
        onClose={this.handleDontEndSessionButtonClick}
        hasCloseCross
      >
        <div className={b()}>
          <div className={b('clock')()} />
          <div className={b('title')()}>{t('AUTH:SESSION-EXPIRATION:SESSION-EXPIRES')}</div>
          <div className={b('timer')()}>
            <div className={b('minutes')()}>{minutesFormatted}</div>:
            <div className={b('seconds')()}>{secondsFormatted}</div>
          </div>
          <div className={b('button')()}>
            <Button size="large" onClick={this.handleDontEndSessionButtonClick}>
              {t('AUTH:SESSION-EXPIRATION:BUTTON-DONT-END-SESSION')}
            </Button>
          </div>
        </div>
      </Modal>
    );
  }

  @bind
  private handleDontEndSessionButtonClick() {
    const { stopTimer, onClose } = this.props;
    stopTimer();
    onClose();
  }
}

export { SessionExpiration, IProps };
export default connect(mapState, mapDispatch)(
  i18nConnect(
    SessionExpiration,
  ),
);
