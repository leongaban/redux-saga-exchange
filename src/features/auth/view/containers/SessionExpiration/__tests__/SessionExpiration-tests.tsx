import * as React from 'react';
import { SessionExpiration, IProps } from '../SessionExpiration';
import { localeProps } from 'shared/helpers/test';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Button } from 'shared/view/elements';

function getShallowComponent(props?: Partial<IProps>) {
  const defaultProps: IProps = {
    ...localeProps,
    isTimerStarted: false,
    timerValue: 0,
    startTimer: jest.fn(),
    stopTimer: jest.fn(),
    onClose: jest.fn(),
    onSessionExpired: jest.fn(),
  };
  return shallow(<SessionExpiration {...defaultProps} {...props} />);
}

describe('Feature Auth', () => {
  describe('container SessionExpiration', () => {
    it('must rendered without crash', () => {
      const component = getShallowComponent();
      expect(toJson(component)).toMatchSnapshot();
    });

    it('endSessionTime: 320, must render timer with 05:20', () => {
      const component = getShallowComponent({ timerValue: 320 });
      expect(toJson(component)).toMatchSnapshot();
    });

    it('click on button must call stopTimer', () => {
      const stopTimer = jest.fn();
      const component = getShallowComponent(({ stopTimer, isTimerStarted: true }));
      expect(stopTimer).toHaveBeenCalledTimes(0);
      component.find(Button).simulate('click');
      expect(stopTimer).toHaveBeenCalledTimes(1);
    });

    it('when component did mount must call startTimer', () => {
      const startTimer = jest.fn();
      const component = getShallowComponent(({ startTimer }));
      component.find(Button).simulate('click');
      expect(startTimer).toHaveBeenCalledTimes(1);
    });
  });
});
