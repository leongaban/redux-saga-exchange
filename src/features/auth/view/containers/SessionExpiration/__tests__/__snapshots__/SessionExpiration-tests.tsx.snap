// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature Auth container SessionExpiration endSessionTime: 320, must render timer with 05:20 1`] = `
<Modal
  hasCloseCross={true}
  isOpen={false}
  onClose={[Function]}
  title="session expiration"
>
  <div
    className="session-expiration__clock"
  />
  <div
    className="session-expiration__title"
  >
    AUTH:SESSION-EXPIRATION:SESSION-EXPIRES
  </div>
  <div
    className="session-expiration__timer"
  >
    <div
      className="session-expiration__minutes"
    >
      05
    </div>
    :
    <div
      className="session-expiration__seconds"
    >
      20
    </div>
  </div>
  <div
    className="session-expiration__button"
  >
    <Button
      onClick={[Function]}
      size="large"
    >
      AUTH:SESSION-EXPIRATION:BUTTON-DONT-END-SESSION
    </Button>
  </div>
</Modal>
`;

exports[`Feature Auth container SessionExpiration must rendered without crash 1`] = `
<Modal
  hasCloseCross={true}
  isOpen={false}
  onClose={[Function]}
  title="session expiration"
>
  <div
    className="session-expiration__clock"
  />
  <div
    className="session-expiration__title"
  >
    AUTH:SESSION-EXPIRATION:SESSION-EXPIRES
  </div>
  <div
    className="session-expiration__timer"
  >
    <div
      className="session-expiration__minutes"
    >
      00
    </div>
    :
    <div
      className="session-expiration__seconds"
    >
      00
    </div>
  </div>
  <div
    className="session-expiration__button"
  >
    <Button
      onClick={[Function]}
      size="large"
    >
      AUTH:SESSION-EXPIRATION:BUTTON-DONT-END-SESSION
    </Button>
  </div>
</Modal>
`;
